### PR TITLE
PI: Don't load entire file into memory when passed file name

### PR DESCRIPTION
**Pull request by mjsir911**
This functionality originally added back in ced2890104dd4633132fe981b7f6bdd67aa17ed8

Reduces memory usage by size of loaded file.

<details>
<summary>Benchmark script</summary>

```
from pypdf import *

filename = '/home/msirabella/tmp/100MB-TESTFILE.ORG.pdf'

writer = PdfWriter(clone_from=filename)

writer.write("out.pdf")
```
</details>

<details>
<summary>Before stats</summary>

```
📏 Total allocations:
	109695

📦 Total memory allocated:
	409.726MB

📊 Histogram of allocation size:
	min: 1.000B
	--------------------------------------------
	< 6.000B   : 40707 ▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇
	< 40.000B  :   229 ▇
	< 256.000B :    33 ▇
	< 1.590KB  : 67394 ▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇
	< 10.104KB :  1060 ▇
	< 64.190KB :   141 ▇
	< 407.789KB:    47 ▇
	< 2.530MB  :    82 ▇
	< 16.072MB :     0 
	<=102.099MB:     2 ▇
	--------------------------------------------
	max: 102.099MB

📂 Allocator type distribution:
	 MALLOC: 107587
	 CALLOC: 1223
	 REALLOC: 865
	 MMAP: 20

🥇 Top 5 largest allocating locations (by size):
	- __init__:./pypdf/_reader.py:315 -> 204.210MB
	- read_from_stream:./pypdf/generic/_data_structures.py:541 -> 101.628MB
	- read_until_regex:./pypdf/_utils.py:233 -> 48.318MB
	- read_object:./pypdf/generic/_data_structures.py:1287 -> 26.012MB
	- _call_with_frames_removed:<frozen importlib._bootstrap>:241 -> 7.360MB

🥇 Top 5 largest allocating locations (by number of allocations):
	- read_until_regex:./pypdf/_utils.py:233 -> 81058
	- read_object:./pypdf/generic/_data_structures.py:1287 -> 23017
	- _call_with_frames_removed:<frozen importlib._bootstrap>:241 -> 2101
	- _compile_bytecode:<frozen importlib._bootstrap_external>:729 -> 988
	- _create_fn:/usr/lib/python3.11/dataclasses.py:433 -> 365
```

</details>

<details>
<summary>After stats</summary>

```
📏 Total allocations:
	109687

📦 Total memory allocated:
	205.521MB

📊 Histogram of allocation size:
	min: 1.000B
	--------------------------------------------
	< 4.000B   : 40707 ▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇
	< 18.000B  :     4 ▇
	< 80.000B  :   227 ▇
	< 348.000B :    39 ▇
	< 1.468KB  : 67239 ▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇
	< 6.341KB  :   737 ▇
	< 27.388KB :   563 ▇
	< 118.297KB:    68 ▇
	< 510.959KB:    21 ▇
	<=2.155MB  :    82 ▇
	--------------------------------------------
	max: 2.155MB

📂 Allocator type distribution:
	 MALLOC: 107587
	 CALLOC: 1218
	 REALLOC: 862
	 MMAP: 20

🥇 Top 5 largest allocating locations (by size):
	- read_from_stream:./pypdf/generic/_data_structures.py:541 -> 101.628MB
	- read_until_regex:./pypdf/_utils.py:233 -> 46.318MB
	- read_object:./pypdf/generic/_data_structures.py:1287 -> 24.012MB
	- _call_with_frames_removed:<frozen importlib._bootstrap>:241 -> 7.356MB
	- _compile_bytecode:<frozen importlib._bootstrap_external>:729 -> 4.844MB

🥇 Top 5 largest allocating locations (by number of allocations):
	- read_until_regex:./pypdf/_utils.py:233 -> 81056
	- read_object:./pypdf/generic/_data_structures.py:1287 -> 23015
	- _call_with_frames_removed:<frozen importlib._bootstrap>:241 -> 2095
	- _compile_bytecode:<frozen importlib._bootstrap_external>:729 -> 989
	- _create_fn:/usr/lib/python3.11/dataclasses.py:433 -> 365
```

</details>

### CODE REVIEW

**Comment by stefan6419846** [comment]
Thanks for the PR. Are the stats correct? You need twice the memory afterwards, thus it would indicate that this is indeed no performance improvement?

And could you please have a look at the failing tests? Your changes lead to new test parallelization issues on Windows as each file can be open only once at each point in time.

File: tests/test_page.py [hunk 1]
```diff
@@ -154,25 +153,25 @@ def test_mediabox_expansion_after_rotation(
...
+        reader_add = PdfReader(pdf_path)
+        page_box = reader_add.pages[0]
+        return page_box, page_base
 
     op = Transformation().scale(2).rotate(45)
 
     # Option 1: The new way
-    page_box1 = deepcopy(page_box)
-    page_base1 = deepcopy(page_base)
+    page_box1, page_base1 = get_pages()
```
[the remark this answers is not in the document]
Comment (mjsir911) [reply tests/test_page.py:169]: I was getting pickle errors due to the fact that it was trying to pickle the pdfreader containing a BufferedIO object

**Comment by mjsir911** [comment]
> Thanks for the PR. Are the stats correct? You need twice the memory afterwards, thus it would indicate that this is indeed no performance improvement?

Sorry, I got the before & after mixed up. fixed

> And could you please have a look at the failing tests? Your changes lead to new test parallelization issues on Windows as each file can be open only once at each point in time.

Yeah, I can do. I'll have a bit more difficulty fixing the windows tests since I don't have a windows box to test on but I'll figure something out.

**Comment by mjsir911** [comment]
Ok yeah the failing tests on windows have been fixed. See the fixed tests to see how this PR introduces potentially incompatible changes to the API ie if a user is writing to a file that's being read from currently.

**Comment by stefan6419846** [comment]
AFAIK the concurrent access issues will only occur on Windows, but I cannot really state how much this would indeed affect real use-cases.

File: pypdf/_reader.py [hunk 2]
```diff
@@ -314,6 +314,7 @@ def __init__(
 
         if isinstance(stream, (str, Path)):
             stream = open(stream, "rb")  #  noqa: SIM115
+            # Wish I could just close stream in __del__ but that fails a test very strangely
```
Comment (stefan6419846) [inline pypdf/_reader.py:317]: Just out of curiosity: Do you have some details about the failure?

Comment (mjsir911) [reply pypdf/_reader.py:317]: Yeah, I'm not sure how much was relevant to drop in the commit but:

when adding a `self.stream.close()` in a `__del__` function, that *does* work most of the time. 

The one test failure I was seeing was in `tests/test_reader.py`, the failing test was `test_get_page_of_encrypted_file` but interestingly this would pass on it's own. I narrowed down the source of the issue to the previous test `test_issue297`'s exception block where the `PdfReader()` initializer was failing (that's what the test is testing for) and the `__del__` block wasn't being called due to the exception happening in the `__init__`. 

It's very possible at some point the objects would be GCd but test failures were happening due to dangling file pointers at the following test.

**Comment by mjsir911** [comment]
> I am not really sure about the fixed tests either - explicitly calling .delete() or even having to close the embedded stream object does not really feel intuitive and maybe even clumsy.

It's even worst than that, unfortunately! I'm not sure what the reference chain is from <Writer> -> <cloned from reader's stream>, but `del writer` seems to unclaim the dangling file pointer.

If it's any consolation, the test updates are kind of an edge case where:
- user is running on windows
- filereader (or `clone_from` transitively) is instantiated via string/Path
- file is acted upon outside of the pdfreader's context (opened again by name) while the pdfreader is still in scope / not garbage collected

Sorry for jumping the gun on calling the tests solved! Still iterating on them.

**Comment by mjsir911** [comment]
> I am not really sure about the fixed tests either - explicitly calling .delete() or even having to close the embedded stream object does not really feel intuitive and maybe even clumsy.
>> It's even worst than that, unfortunately! I'm not sure what the reference chain is from -> <cloned from reader's stream>, but del writer seems to unclaim the dangling file pointer.

I could potentially add a `.close()` or something to `PdfReader` which would at least make this process explicit. I would still be unsure how to propogate that to `PdfWriter`'s API thoug.

File: tests/test_page.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (mjsir911) [reply tests/test_page.py:169]: @stefan6419846 do we support deepcopying objects? The coverage report has lost a line of coverage because the deepcopy code was only being covered incidentally and I don't think will ever work anymore.

File: pypdf/_reader.py [hunk 3]
```diff
@@ -342,6 +346,9 @@ def __init__(
         elif password is not None:
             raise PdfReadError("Not encrypted file")
 
+    def close(self):
+        self.stream.close()
+
```
Comment (pubpub-zz) [inline pypdf/_reader.py:351]: it would be great to add comments for documentation

File: tests/test_page.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (pubpub-zz) [reply tests/test_page.py:169]: deep copy should be avoided. By their own there are meaning less in pypdf because you need to attach it to PdfWriter. clone should be used instead

File: pypdf/_reader.py (hunk 2 above)
Comment (mjsir911) [reply pypdf/_reader.py:317]: I'm going to add this to the commit

**Comment by mjsir911** [comment]
I don't want this merged as it currently is, calling garbage collection manually in tests feels yucky.

**Comment by pubpub-zz** [comment]
> It's even worst than that, unfortunately! I'm not sure what the reference chain is from «Writer» -> «» «cloned from reader's stream», but `del writer` seems to unclaim the dangling file pointer.

when you call `.clone_document_from_reader()` or `append(pages)`, you clone all objects from PdfReader(). during this process we need to keep connection between the writer's objects and the reader's object in order to keep parents links for example.
When you have finished your work, or when you need to append a new set of pages detached from the PdfReader, you have to call
`writer.reset_translation()`.

File: pypdf/_reader.py [hunk 4]
```diff
@@ -349,6 +353,20 @@ def close(self) -> None:
...
+        """Use PdfReader as context manager"""
+        return self
+
+    def __exit__(
+        self,
+        exc_type: Optional[Type[BaseException]],
+        exc: Optional[BaseException],
+        traceback: Optional[TracebackType],
+    ) -> None:
+        """Write data to the fileobj."""
```
Comment (pubpub-zz) [inline pypdf/_reader.py:366]: the comment does not seem correct

File: pypdf/_reader.py [hunk 5]
```diff
@@ -349,6 +353,20 @@ def close(self) -> None:
...
+        if self._we_opened:
+            self.close()
```
Comment (pubpub-zz) [inline pypdf/_reader.py:368]: can you add test to maintain test coverage please

Comment (stefan6419846) [reply pypdf/_reader.py:368]: Additionally, I would prefer something like `_opened_automatically` instead of `_we_opened` to sound "more generic".

**Comment by mjsir911** [comment]
I should also add using `PdfReader` as a contextmanager in some documentation somewhere

File: pypdf/generic/_base.py [hunk 6]
```diff
@@ -283,9 +283,6 @@ def indirect_reference(self) -> "IndirectObject":  # type: ignore[override]
     def get_object(self) -> Optional["PdfObject"]:
         return self.pdf.get_object(self)
 
-    def __deepcopy__(self, memo: Any) -> "IndirectObject":
-        return IndirectObject(self.idnum, self.generation, self.pdf)
-
```
Comment (pubpub-zz) [inline pypdf/generic/_base.py:288]: I'm not so found about removing __deepcopy__ : some people may use it this could be considered as a regression. If we really want to remove it we shall use the depredication process

**Comment by pubpub-zz** [comment]
> I should also add using `PdfReader` as a contextmanager in some documentation somewhere

have you also been able to advance in your proposal ?

**Comment by mjsir911** [comment]
> have you also been able to advance in your proposal ?

Hi, sorry, I've been taking a break from things due to mental health but plan to be back on them sometime later next month. Moving this back to draft for now.

File: pypdf/generic/_base.py (hunk 6 above)
Comment (mjsir911) [reply pypdf/generic/_base.py:288]: @pubpub-zz deprecating doesn't really make sense because with this change no objects will ever be deep-copyable, they will always have a reference to a file pointer that can't be pickleable. 

The only reason deep copies work now is because the entire source PDF bytestring gets copied over with them, and that only happens when a filename is passed, deepcopying has never worked with a passed file pointer.

Comment (mjsir911) [reply pypdf/generic/_base.py:288]: The only way deprecation would work is if you deprecated it in lieu of this PR and then merged these changes in at a later date

Comment (pubpub-zz) [reply pypdf/generic/_base.py:288]: If you leave `__deepcopy__`  is there an error ?

Comment (mjsir911) [reply pypdf/generic/_base.py:288]: if I leave `__deepcopy__` with the associated covered tests there is an error, yes.

<details>
<summary> 

`TypeError: cannot pickle '_io.FileIO' object`

</summary>

```
    res = hook_impl.function(*args)
          ^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/msirabella/fork/pypdf/venv/lib/python3.11/site-packages/_pytest/python.py", line 195, in pytest_pyfunc_call
    result = testfunction(**testargs)
             ^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/msirabella/fork/pypdf/tests/test_page.py", line 168, in test_transformation_equivalence
    page_box1 = deepcopy(page_box)
                ^^^^^^^^^^^^^^^^^^
  File "/usr/lib/python3.11/copy.py", line 172, in deepcopy
    y = _reconstruct(x, memo, *rv)
        ^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/lib/python3.11/copy.py", line 271, in _reconstruct
    state = deepcopy(state, memo)
            ^^^^^^^^^^^^^^^^^^^^^
  File "/usr/lib/python3.11/copy.py", line 146, in deepcopy
    y = copier(x, memo)
        ^^^^^^^^^^^^^^^
  File "/usr/lib/python3.11/copy.py", line 231, in _deepcopy_dict
    y[deepcopy(key, memo)] = deepcopy(value, memo)
                             ^^^^^^^^^^^^^^^^^^^^^
  File "/usr/lib/python3.11/copy.py", line 172, in deepcopy
    y = _reconstruct(x, memo, *rv)
        ^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/lib/python3.11/copy.py", line 271, in _reconstruct
    state = deepcopy(state, memo)
            ^^^^^^^^^^^^^^^^^^^^^
  File "/usr/lib/python3.11/copy.py", line 146, in deepcopy
    y = copier(x, memo)
        ^^^^^^^^^^^^^^^
  File "/usr/lib/python3.11/copy.py", line 231, in _deepcopy_dict
    y[deepcopy(key, memo)] = deepcopy(value, memo)
                             ^^^^^^^^^^^^^^^^^^^^^
  File "/usr/lib/python3.11/copy.py", line 161, in deepcopy
    rv = reductor(4)
         ^^^^^^^^^^^
TypeError: cannot pickle '_io.FileIO' object
```
</details>

**Comment by stefan6419846** [comment]
What is the current state of this PR?
